File: 백준/Silver/9012. 괄호/괄호.py
#9012
import sys

class Stack:
    def __init__(self):
        self.stack = []
        
    def push(self, bracket):
        self.stack.append(bracket)
    
    def pop(self):
        if not self.stack:
            return False #스택이 비어있는 경우
        else:
            return self.stack.pop()
        
    def is_empty(self):
        return not bool(self.stack)
    #비어 있으면 True, 아니면 False
    
def check_vps(string):
    stack = Stack()
    for s in string:
        if s == '(':
            stack.push(s)
        elif s == ')':
            if stack.pop() == False:
                return "NO" #비어 있으면 NO
    if stack.is_empty():
        return "YES" #True면 YES
    else:
        return "NO"
            
n = int(sys.stdin.readline())
for i in range(n):
    string = sys.stdin.readline().rstrip()
    print(check_vps(string))

File: 백준/Silver/9012. 괄호/괄호.cc
//9012
// 괄호 검사(VPS)

#include <iostream>
#include <stack>
#include <string>

using namespace std;

int main() {
    int n;
    cin >> n;

    while (n > 0) {
        n--;

        string str;
        cin >> str;

        stack<char> s;
        string ans = "YES";

        for (int i = 0; i < (int)str.size(); i++) {
            if (str[i] == '(') {
                s.push(str[i]);
            }
            else if (!s.empty() && str[i] == ')' && s.top() == '('){
                s.pop();
                }
            else {
                ans = "NO";
                break;
            }
        }

        if (!s.empty())
            ans = "NO";

        cout << ans << endl;
    }

    return 0;
}